### ISSUE TITLE
[Bug] openmp in custom ops is very slow

### ISSUE THREAD

**Issue by autumn110** [openvinotoolkit/openvino#16503]
##### System information (version)

- OpenVINO => built from master
- Operating System / Platform => WSL2
- Compiler => gcc 7.5
- Problem classification: custom ops with openmp

##### Detailed description
I've written a custom op which I want to improve the performance with openmp, but find it very slow. I've did some experiments that compared the same test function with/without openvino, the result shows that in evaluate function, the performance is poor.

test function in openvino evaluate function, the running time:
![image](https://user-images.githubusercontent.com/124126622/227144082-d6a899ba-385d-4960-88ba-bac5869b4790.png)

test function in a separate main function without openvino, the running time:
![image](https://user-images.githubusercontent.com/124126622/227144304-893127c1-e86b-49ce-b2e3-36e21cb0b5d7.png)

##### Steps to reproduce

use the test code, build test and test1 targets.
run test for the openvino version
run test1 without openvino
[openmp_test.zip](https://github.com/openvinotoolkit/openvino/files/11048499/openmp_test.zip)

##### Issue submission checklist

 - [x] I report the issue, it's not a question
 - [x] I checked the problem with documentation, FAQ, open issues, Stack Overflow, etc and have not found solution
 - [x] There is reproducer code and related data files: images, videos, models, etc.

**Comment by autumn110** [comment]
Hi @ilya-lavrenov , thanks for your advice. It works after switching openvino with openmp version. But I found another weird issue. It seems like the extension runs in the same speed w/wo openmp directives.

when I print the following code in the omp for loop 
```
        std::cout << "Num threads = " << omp_get_num_threads() << std::endl;
        std::cout << "Max threads = " << omp_get_max_threads() << std::endl;
        std::cout << "In parallel? " << omp_in_parallel() << std::endl;
```
the results are always like
```
Num threads = 1
Max threads = 8
In parallel? 0
```

And another weird thing is that the performance of extension looks like the same as the same function with omp in separate main target without openvino. So I guess the extension has already been improved by certain factor. I've no idea how to understand these cases.

**Comment by wangleis** [comment]
@autumn110 Could you share the capture of htop during inference to check if application run on multi-threading or single threading during?

**Comment by autumn110** [comment]
hi @wangleis , after i did some test, maybe the issue is not related to openvino but intel openmp. 

It looks like linking libiomp causes the single threading running. here's some screenshots

**with libiomp**, weird single threading
![Dingtalk_20230324184708](https://user-images.githubusercontent.com/124126622/227503353-09ccd028-0592-469e-ad24-e800999df1ac.jpg)
![Dingtalk_20230324184506](https://user-images.githubusercontent.com/124126622/227503412-449893d2-097d-4d0e-9c95-af1167265c5d.jpg)

**with libgomp**, normal multi-threading
![Dingtalk_20230324184724](https://user-images.githubusercontent.com/124126622/227503549-ef8acd6c-0fb9-469d-a32a-bdf2b004a95f.jpg)
![Dingtalk_20230324184524](https://user-images.githubusercontent.com/124126622/227503594-20ef4f84-6ad8-469d-8a98-ea843dc4b3b3.jpg)

No sure if i'm doing the right thing, but the result is very strange with libiomp

here is my test code

[iomp_gomp.zip](https://github.com/openvinotoolkit/openvino/files/11061331/iomp_gomp.zip)

**Comment by autumn110** [comment]
I'm sorry for my mistake. I find when linking the libiomp5.so, -fopenmp flag is not added, that's the reason why openmp not working in openvino. By the way, which way do you recommend, using openvino and extension with default tbb or openmp?

**Comment by ilya-lavrenov** [comment]
We prefer to use TBB and this is default threading framework for OpenVINO.

**Comment by autumn110** [comment]
hi @ilya-lavrenov , I tested my function both with openmp and tbb, find that tbb is much slower than omp. As I'm not an expert of threading model, have no idea what is happening here. Could you give me some advise? 

Another issue I observed is that when I run the function with openmp multiple times. it runs faster and faster each time. for the first time running, I know it loads the dynamic lib causing much time, but why the second time still causes a lot than later as follows.  It's always like that
![image](https://user-images.githubusercontent.com/124126622/227877466-639d10b9-78b4-48ce-ada8-80f80935e761.png)



here is my test code.
[tbb_test.zip](https://github.com/openvinotoolkit/openvino/files/11076108/tbb_test.zip)

**Comment by wangleis** [comment]
@autumn110 Do you mean OpenVINO build based on OpenMP is slower than OpenVINO build based on TBB?

If yes, may I know what OS version and TBB version you are using now?

**Comment by wangleis** [comment]
@autumn110 You can run `benchmark_app -m model -d CPU -hint throughput` on both OpenMP based OpenVINO and TBB based OpenVINO to check if it is system issue. We do not observe much slower performance on TBB.

Since it is not OpenVINO related issue, we will close it if there is no more question.

**Comment by wangleis** [comment]
Close this issue. Please feel free to reopen or create new issue if additional assistance is needed.

**Comment by autumn110** [comment]
hi @wangleis , sorry for disturbing again. I encountered a new issue with extension written by openmp. 

When I use an old master version saying fc98454174, the omp(only use `#pragma omp parallel for`) works very well as expected. However when I pull the lastest master, the loop could only use single thread.  After I manually set the thread num like `omp_set_num_threads(8)`, everything becomes normal again.

From some instructions, the default thread number used by `#pragma omp parallel for` should be the same as cpu thread number. So I'm wondering why the default thread number becomes 1 in the latest master of openvino. Could you give me some advice?

Thanks in advance

**Comment by wangleis** [comment]
@autumn110 Thanks for your information. Will verify and fix the issue on master branch.

**Comment by wangleis** [comment]
@autumn110 OpenMP on master branch has been fixed. Please try it. Thanks.

**Comment by autumn110** [comment]
@wangleis  Current master branch works very well. Thanks a lot!
